File: LockServ/Lock Server.java
import java.util;
import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;

/*
	1. central lock Server node
	
	2. lock Agent runs on each client node
		--> each client node accesses the API through the Agent
	
	3. each lock Agent communicats over the network w/ the
		central lock Server

	4. Agent requests & releases the lock with
		--> LockMsg
		--> UnlockMsg

	5. Server sends a msg to notify Agent when it has received the lock
		--> GrantMsg
	
	** Mutual Exclusion **

		= only 1 node can hold lock at a time
*/

public class ActorNode extends AbstractActor{
	private final LoggingAdapter log = Logging.getLogger(context().system(), this);
	// "name" of each node
	public boolean isAgent;
	// each Agent indicates when holding lock
	public boolean holdsLock;
	// Server maintains a queue of Agent nodes
	List<ActorNode> queue;

	public ActorNode(String args){
		// matches types of messages actor can accept
		receive(ReceiveBuilder.
      		match(String.class, s -> {
        		log.info("Received String message: {}", s);
      		}).
      		matchAny(o -> log.info("received unknown message")).build()
    	);
		/* Don't think I need InitState since I initialize here */
    	holdsLock = false;
		queue = new ArrayList<ActorNode>();
		if(isAgent(args))
			isAgent = 1;
		else
			isAgent = 0;
	}

	void public State(Node args){
		// i. currently holding lock
		if(queue.indexOf(args) == 0)
			args.holdsLock = true;
		// ii. add to queue if not empty
		else{
			if(!queue.contains(args))
				queue.add(args);
		}
	}

	void HandleInp(Node args, String input){
		if(isAgent(args)){
			if(input.equals("LockMsg"))
				// TO-DO: tell Server to Lock
			else{
				if(args.holdsLock)
					args.holdsLock = false;
				// TO-DO: tell Server to Unlock
			}
		}
	}

	void HandleMsg(Node args, Node src, String input){
		// i. Agent
		if(isAgent(args)){
			// TO-DO: tell Server to Grant Msg
			args.holdsLock = true;
		}
		// ii. Server
		else{
			// LOCK
			if(input.equals("LockMsg")){
				// If lock not held, immediately grant lock
				if(s.queue.size() == 0)
					// TO-DO: tell Server to Grant Msg immediately
				s.queue.add(src);
			}
			// UNLOCK
			else{
				// Send head of queue to back -- no longer holds lock
				Node temp = s.queue.get(0);
				s.queue.remove(0);
				s.queue.add(temp);
				// Grant Lock to next waiting agent -- if any!
				if(s.queue.size() > 0)
					// TO-DO: GrantMsg to next waiting agent
			}
		}
	}
	
	boolean isAgent(String args){
		if(args.equals("Agent") || args.equals("agent"))
			return true;
		return false;
	}
}

File: LockServ/src/main/scala/lockserver.scala
package lockserv
import akka.actor.Actor
import scala.collection.mutable.Queue

/* - let Agent class take an output ActorRef to send outputs to */

/*
 *  [ Name :=   Server | Agent(n) ]
 *
 *  Name lists the names of nodes in the system.
 *
 *  There is 
 *
 *    1) a single Server node 
 *    2) an arbitrary # of Agents
 *
 *
 *  [ State ]
 *
 *  Defines the state maintained at each node.
 *  Node's Name determines the data maintained locally at that node.
 *
 *  Server := maintains a [queue] of Agent nodes
 *    - initially empty
 *    - head: Agent holding the lock
 *    - rest: Agents waiting to acquire the lock
 *
 *  Agent := maintains a [boolean]
 *    - initially false
 *    - true when Agent holds the lock
 */

abstract class Name extends Actor{}

class Server extends Name{
  val state :Queue[Agent] = new Queue()
  def receive = {
    case LockMsg => {
      /* Does locking just consist of updating the state? */
      if (this.head().getState() == false)
        this.head().setState(true)
    }
    case UnlockMsg => {
      /* Since I already set the state to false,
       Do I need to shiftHead here? */
    }
    case GrantedMsg(a) =>{
      /* Same concerns as above! */
      if(a.getState() == 0)
        a.setState(true)
    }
  } 
  def queueEmpty() : Boolean = {
    return state.isEmpty
  }
  def addAgent(a: Agent) : Unit = {
    state += (a)
  }
  def shiftHead = {
    val oldhead = state.dequeue
    state += (oldhead)
  }
  def head() : Agent = {
    return state.front
  }
}

class Agent(server: ActorRef) extends Name{
  var state : Boolean = false
  var server : Server = server

  def receieve = {}
  def getState() : Boolean = {
    return state
  }
  def setState(b: Boolean) : Unit = {
    state = b
  }
  def getServer() : Server = {
    return server
  }
}


/*  
 *  [ Input := Lock | Unlock ]
 *
 *  External I/O messgaes exchanged btwn Agent & other local processes on its node
 */
trait Input
case object Lock extends Input
case object Unlock extends Input

/*  
 *  [ Msg := LockMsg | UnlockMsg | GrantedMsg ]
 *
 *  Network messages exchanged between Agents and the central Server
 */
abstract class Msg
case object LockMsg extends Msg
case object UnlockMsg extends Msg
case class GrantedMsg(src: Agent) extends Msg

object Handler{
  
  def HandleInput(n: Name, inp: Input) : Unit = {
    n match {
      case s: Server => {}
      case a: Agent => {
        val t = n.asInstanceOf[Agent]
        inp match {
          case Lock =>{
            t.getServer() ! LockMsg
          }
          case Unlock => {
            if(t.getState() == true)
              t.setState(false)
            t.getServer() ! UnlockMsg
          }
        }
      } 
    }
  }

  def HandleMsg(n: Name, src: Name, msg: Msg) : Unit = {
    n match {
      case s: Server => {
        val t1 = n.asInstanceOf[Server]
        msg match {
          case LockMsg => {
            if(t1.queueEmpty())
              t1 ! GrantedMsg(src.asInstanceof[Agent]) // grant lock to src if lock not held
            t1.addAgent(src)
          }
          case UnlockMsg => {
            t1.shiftHead
            if(!t1.queueEmpty())
              t1 ! GrantedMsg(t1.head) // grant lock to new head
          }
        } 
      }
      case a: Agent => {
        val t2 = n.asInstanceOf[Agent]
        msg match {
          case GrantedMsg(_) => {
            t2.setState(true)
            // output Granted
          }
        } 
      }
    }
  }
}
